### PR TITLE
Fix RoutingDataCache always requiring full refresh for current state.

### PR DESCRIPTION
**Pull request by pkuwm**
### Issues

- [x]  My PR addresses the following Helix issues and references them in the PR title:
#483 Fix RoutingDataCache always does full refresh for current state.

### Description

- [x]  Here are some details about my PR, including screenshots of any UI changes:
(Write a concise description including what, why, how)

#483 Fix RoutingDataCache always does full refresh for current state.
If a current state changes, propertyDataMap doesn't mark current state type not to do full refresh next time. This causes current state refresh to always do full refresh. And this increases snapshot read latency. So we need to mark propertyDataMap's current state type to false, NOT requiring full refresh next time.

And for live instance changes, we need to mark propertyDataMap's current state to true, requiring current state full refresh next time.

### Tests

- [x]  The following tests are written for this issue:
TestRoutingDataCache.testCurrentStatesSelectiveUpdate

- [x]  The following is the result of the "mvn test" command on the appropriate module:
(Copy & paste the result of "mvn test")

[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   TestAlertingRebalancerFailure.testTagSetIncorrect:174->checkResourceBestPossibleCalFailureState:310 expected:<true> but was:<false>
[ERROR]   TestPartitionLevelTransitionConstraint.test:173 expected:<true> but was:<false>
[ERROR]   TestZkConnectionLost.testLostZkConnection:140 » Helix Workflow "testLostZkConn...
[ERROR]   TestDeleteJobFromJobQueue.testForceDeleteJobFromJobQueue:75 » Helix Failed to ...
[ERROR]   TestTaskPerformanceMetrics.testTaskPerformanceMetrics:118 expected:<true> but was:<false>
[INFO]
[ERROR] Tests run: 882, Failures: 5, Errors: 0, Skipped: 1
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  57:51 min
[INFO] Finished at: 2019-09-18T17:51:09-07:00

[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.925 s - in org.apache.helix.integration.task.TestDeleteJobFromJobQueue
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.178 s
[INFO] Finished at: 2019-09-19T09:27:36-07:00

[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.92 s - in org.apache.helix.integration.TestAlertingRebalancerFailure
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  8.357 s
[INFO] Finished at: 2019-09-19T09:26:43-07:00

[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 16.092 s - in org.apache.helix.monitoring.mbeans.TestTaskPerformanceMetrics
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  20.176 s
[INFO] Finished at: 2019-09-19T09:28:14-07:00

[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.066 s - in org.apache.helix.integration.TestPartitionLevelTransitionConstraint
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10.440 s
[INFO] Finished at: 2019-09-19T09:29:22-07:00
[INFO] ------------------------------------------------------------------------

### Commits

- [x]  My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "How to write a good git commit message":
Subject is separated from body by a blank line
Subject is limited to 50 characters (not including Jira issue reference)
Subject does not end with a period
Subject uses the imperative mood ("add", not "adding")
Body wraps at 72 characters
Body explains "what" and "why", not "how"
Documentation

  In case of new functionality, my PR adds documentation in the following wiki page:
(Link the GitHub wiki you added)

### Code Quality

- [x]  My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java [hunk 1]
```diff
@@ -113,6 +113,7 @@ public void refresh(HelixDataAccessor accessor) {
     if (_propertyDataChangedMap.get(HelixConstants.ChangeType.LIVE_INSTANCE)) {
       long start = System.currentTimeMillis();
       _propertyDataChangedMap.put(HelixConstants.ChangeType.LIVE_INSTANCE, Boolean.valueOf(false));
+      _propertyDataChangedMap.put(HelixConstants.ChangeType.CURRENT_STATE, true);
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: Keep the style the same, please.

**Comment by pkuwm** [comment]
> Was this a bug? Also could we add a quick unit test for this?

Unit test is what I am working on now. Will add a unit test later soon.

**Comment by narendly** [comment]
@pkuwm That still doesn't answer my question. Was this originally a bug in the routing cache? Or are you having to fix this as a result of some recent change?

**Comment by pkuwm** [comment]
> @pkuwm That still doesn't answer my question. Was this originally a bug in the routing cache? Or are you having to fix this as a result of some recent change?

From my understanding, it originally was done on purpose as current state refresh uses selective update. I wouldn't say it was a bug in the routing cache as it did not affect correctness. But it actually did not reset _propertyDataChangeMap for current states cache. Each time current states cache would be refreshed. So performance may be impacted. Resetting _propertyDataChangeMap for current state would help the performance.

**Comment by dasahcc** [comment]
Please update comments as merge steps wiki showed.

**Comment by pkuwm** [comment]
Going to add a test before merging it.

**Comment by narendly** [comment]
@pkuwm Any update on this PR? :)

**Comment by pkuwm** [comment]
> @pkuwm Any update on this PR? :)

I am on track. Will complete a unit test for this after metrics work is done.

**Comment by pkuwm** [comment]
@dasahcc @jiajunwang @narendly Unit test is ready for review. Thanks.

**Comment by pkuwm** [comment]
This PR is ready to be merged, approved by @dasahcc @jiajunwang 

commit message:
```
Fix #483  RoutingDataCache always does full refresh for current state.
If a current state changes, propertyDataMap doesn't mark current state type not to do refresh. 
This causes current state cache to always refresh. And this increases snapshot read latency. 
So we need to mark propertyDataMap's current state type to false, NOT requiring full refresh next time.

And for live instance changes, we need to mark propertyDataMap's current state to true, requiring current state cache refresh next time.

Change list:
1. mark current state cache to also refresh when live instance change happens.
2. if current state cache is already refreshed, no need to refresh next time if no live instance change.
3. add unit test TestRoutingDataCache.testCurrentStatesSelectiveUpdate

```

**Comment by pkuwm** [comment]
@dasahcc Can you help merge this PR? Thanks.

**Pull request merged**